sw/host/opentitanlib/src/spiflash/sfdp.rs

```diff
@@ -0,0 +1,529 @@
+                wait_states: p.wait_states_144(),
+                mode_bits: p.mode_bits_144(),
+                opcode: p.read_opcode_144(),
+            }
+        }
+        if jedec.support_fast_read_114 {
+            jedec.param_114 = FastReadParam {
+                wait_states: p.wait_states_114(),
+                mode_bits: p.mode_bits_114(),
+                opcode: p.read_opcode_114(),
+            }
+        }
+        if jedec.support_fast_read_222 {
+            jedec.param_222 = FastReadParam {
+                wait_states: p.wait_states_222(),
+                mode_bits: p.mode_bits_222(),
+                opcode: p.read_opcode_222(),
+            }
+        }
+        if jedec.support_fast_read_444 {
+            jedec.param_444 = FastReadParam {
+                wait_states: p.wait_states_444(),
+                mode_bits: p.mode_bits_444(),
+                opcode: p.read_opcode_444(),
+            }
+        }
+        Ok(jedec)
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct UnknownParams {
+    pub data: Vec<u32>,
+}
+
+impl TryFrom<&[u8]> for UnknownParams {
+    type Error = Error;
+    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
+        let mut rdr = std::io::Cursor::new(buf);
+        let mut data = vec![0u32; buf.len() / 4];
+        rdr.read_u32_into::<LittleEndian>(&mut data)
+            .map_err(|e| Error::Io(e))?;
+        Ok(UnknownParams { data })
+    }
+}
+
+/// The `Sfdp` structure represents the decoded SFDP table.
+#[derive(Debug, Serialize)]
+pub struct Sfdp {
+    pub header: SfdpHeader,
+    pub phdr: Vec<SfdpPhdr>,
+    pub jedec: JedecParams,
+    pub params: Vec<UnknownParams>,
+}
+
+impl Sfdp {
+    /// Given an initial SFDP buffer calculate the number of bytes needed for
+    /// the entire SFDP.
+    pub fn length_required(buf: &[u8]) -> Result<usize, Error> {
+        if buf.len() < 256 {
+            // Technically, we could read out the first 8-bytes, then
+            // figure out how many phdrs there are, read out that number
+            // times 8-byte chunks, then find the max extent of the pointed-to
+            // headers and perform the calculation in the else-arm below.
+            //
+            // We punt and read out the first 256 bytes, which is often more
+            // than enough.  In the event it isn't, we assume the first 256
+            // bytes will contain all of the phdrs (it will) and perform
+            // the max-extent calculation below.
+            Ok(256)
+        } else {
+            let header = SfdpHeader::try_from(&buf[0..8])?;
+            let mut len = 0;
+            for i in 0..=header.nph {
+                let start = (8 + i * 8) as usize;
+                let end = start + 8;
+                let phdr = SfdpPhdr::try_from(&buf[start..end])?;
+                len = std::cmp::max(len, phdr.offset + (phdr.dwords * 4) as u32);
+                log::debug!("computed sfdp len = {}", len);
+            }
```

Comment:
You don't have to do this but...

```rust
(0..=header.nph).map(|i| ...; len).max()
```

Reply:
If I move the length calculation into the lambda, I can't use `?` to short-circuit out on errors from the `SfdpPhdr` try_from conversion (or can I?).

Reply:
There is no `try_map`. Very irritating. You would need to use `try_fold`, which is an unnecessary complication IMO.

Reply:
Honestly, this just seem cleaner as a loop to me.